cel/cel_test.go

```diff
@@ -1263,3 +1266,48 @@ func TestResidualAst_Modified(t *testing.T) {
 		}
 	}
 }
+
+type envExpected struct {
+	declarations []*exprpb.Decl
+}
+
+func TestDeclareContextProto(t *testing.T) {
+	descriptor := new(tpvpb.TestProtoValue).ProtoReflect().Descriptor()
+	option := DeclareContextProto(descriptor)
+	got, err := NewEnv(option)
+	if err != nil {
+		t.Fatalf("NewEnv(DeclareContextProto(%v)) failed: %s", descriptor, err)
+	}
+	wantDeclarations := []*exprpb.Decl{
+		decls.NewVar("single_int64", decls.Int),
+		decls.NewVar("single_double", decls.Double),
+		decls.NewVar("single_bool", decls.Bool),
+		decls.NewVar("single_string", decls.String),
+		decls.NewVar("single_nested_message", decls.NewObjectType("cel.testdata.TestProtoValue.NestedMessage")),
+		decls.NewVar("single_nested_enum", decls.Int),
+		decls.NewVar("single_duration", decls.Duration),
+		decls.NewVar("single_timestamp", decls.Timestamp),
+		decls.NewVar("single_any", decls.Any),
+		decls.NewVar("repeated_int32", decls.NewListType(decls.Int)),
+		decls.NewVar("map_int64_string", decls.NewMapType(decls.Int, decls.String)),
+	}
+
+	want, err := NewEnv(Declarations(wantDeclarations...))
+	if err != nil {
+		t.Fatalf("test setup failed: %s", err)
+	}
+	cmp.AllowUnexported()
```

Comment:
Maybe instead of testing the internal object structure of the `Env`, you could instead type-check and evaluate an expression which contains type tests these variables:

```cel
type(single_int64) == int && type(single_double) == double && ...
```

Note that `list` and `map` types are not parameterized at runtime and so `type(map_int64_string) == map` is as much information as the type-checker would validate.

Alternatively, you could attempt to test an expression which performs equality and then you only need to use the type-checker to validate behavior:

```cel
single_int64 == 1 && map_int64_string == {42: "answer to everything"} && ...
```

Reply:
Simplified with second suggestion.